### PR TITLE
FFF-103: Message builder pattern

### PR DESCRIPTION
**Pull request by adlrocha**
Addresses https://github.com/filecoin-project/go-f3/issues/103. Continuation (and depends on) https://github.com/filecoin-project/go-f3/pull/259

This PR:

- Removes the id from gpbft.Participant.
- Changes the signature and moves `MarshalPayloadForSigning` out of `Signature` so ti can be used in `MessageBuilder`. Without this, the serialization performed by the verifier and the message builder didn't match in tests  leading to a nasty bug to find.  
- Uses message builder to allow the signature of message by several hosts.

### CODE REVIEW

File: gpbft/participant_test.go
Comment (masih) [inline gpbft/participant_test.go:89]: I would leave a comment explaining the rationale for `Maybe()`

File: gpbft/vrf.go
Comment (masih) [inline gpbft/vrf.go:17]: Discussion point for later date: once the number of arguments to a function go above 3 I would start to question what we might be missing in the API/interface design.

Hopefully once the implementation settles we would revisit.

**Review by masih — APPROVED** [review]
This is a PR to an open PR. Once flattened I'll give it another look. For now no blockers from me 👍

File: gpbft/message_builder.go
Comment (adlrocha) [reply gpbft/message_builder.go:53]: They definitely shouldn't be exported. I copy-pasted from Kuba's PR and didn't even stop twice to think about it. I will address it in my next commit. Good catch. Thanks!

File: gpbft/message_builder.go
Comment (adlrocha) [reply gpbft/message_builder.go:17]: Same as above

File: gpbft/message_builder.go
Comment (adlrocha) [reply gpbft/message_builder.go:63]: As Kuba mentioned,  he was using this convention to signal unmutable functions, i.e. functions that don't modify the object internally. I kind of like it, so I will keep it like that.

File: gpbft/message_builder.go
Comment (adlrocha) [reply gpbft/message_builder.go:161]: I refactor it a bit to make `sign` and `build` private functions of `SignatureBuilder` because right now they are just internally used by `MessageBuilder`

File: gpbft/message_builder.go
Comment (adlrocha) [reply gpbft/message_builder.go:145]: Same as above, unless strongly opposed, I am using Kuba's convention of using pointers only for mutable methods.

File: gpbft/message_builder_test.go
Comment (adlrocha) [reply gpbft/message_builder_test.go]: fixed

File: gpbft/api.go
Comment (adlrocha) [reply gpbft/api.go:62]: 1. is addressed, but for 2. I thought the same but all my attempts felt a bit dirty. Maybe I am missing some easy way, any proposal? Thanks!

**Comment by adlrocha** [comment]
@masih, I think I addressed all of your comments. Let me know if there is any outstanding one. If this is not the case, we can do what we discussed sync of merging this one to the chained PR and have a final review of the whole chain in a single PR. 

Unless @Kubuxu, or any one else wants to review this one before merging to the dependent branch, in that case please let me know before tomorrow. Thanks :pray:

**Comment by masih** [comment]
Thanks @adlrocha please feel free to merge this once the test failure is fixed. We can follow up on any outstanding items in the flattened PR 🙏

**Review by anorth — APPROVED** [review]
Thanks. 

Please check that with tracing enabled, the log from a basic simulation test looks similar to before this change. Otherwise LGTM.

I'd prefer you rebase this against main once upstream has merged, rather than combined into one PR, if possible.

**Review by masih — APPROVED** [review]

**Comment by anorth** [comment]
This did end up breaking the participant ID in trace logs. Fixed in #313.

**Pull request merged**
